README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# gridbr: easy access to the Brazilian statistical grid with R <a href='https://lucunha.com/gridbr'><img src='man/figures/gridbr.png' align="right" height="138.5" /></a>

<!-- badges: start -->

[![R-CMD-check](https://github.com/luisfelipebr/gridbr/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/luisfelipebr/gridbr/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

**The goal of ‘gridbr’ is to provide easy access to the [Brazilian
Statistical
Grid](https://mapasinterativos.ibge.gov.br/grade/default.html)**,
published by the Brazilian Institute of Geography and Statistics (IBGE).
It builds upon the development made in the project [IBGE Statistical
Grid in Compact Representation](https://github.com/osm-codes/BR_IBGE).

## Installation

You can install the development version of *gridbr* from this GitHub
repository with `devtools`:

``` r
devtools::install_github("luisfelipebr/gridbr")
```

## Usage

### Setup

``` r
library(gridbr)
library(sf)
```

In the following examples, the Brazilian archipelago [Fernando de
Noronha](https://en.wikipedia.org/wiki/Fernando_de_Noronha) is used as
aoi. It can be download with the package
[geobr](https://ipeagit.github.io/geobr/) by using its municipality
code.

``` r
library(geobr)
```

``` r
aoi <- read_municipality(2605459, showProgress = FALSE)
```

``` r
plot(st_geometry(aoi))
```

<img src="man/figures/README-figure1-1.png" width="100%" />

### gridbr_download(input, cellsize)

**`gridbr_download()`** is the main function available in the package.
It builds the original statistical grid published by IBGE and merge it
with downloaded 2010 population census data. The user must provide the
area of interest (aoi) as a geospatial
[sf](https://r-spatial.github.io/sf/index.html) object.

Two arguments are mandatory: `input` and `cellsize`.

``` r
aoi_grid <- gridbr_download(input = aoi, 
                            cellsize = "1KM")
```

``` r
head(aoi_grid)
#> Simple feature collection with 6 features and 5 fields
#> Geometry type: POLYGON
#> Dimension:     XY
#> Bounding box:  xmin: -32.48123 ymin: -3.890802 xmax: -32.46804 ymax: -3.836282
#> Geodetic CRS:  SIRGAS 2000
#>               id MASC FEM POP DOM_OCU                       geometry
#> 1 1KME7376N10816    0   0   0       0 POLYGON ((-32.47706 -3.8908...
#> 2 1KME7376N10817    0   0   0       0 POLYGON ((-32.47775 -3.8818...
#> 3 1KME7376N10818    0   0   0       0 POLYGON ((-32.47845 -3.8728...
#> 4 1KME7376N10819    0   0   0       0 POLYGON ((-32.47914 -3.8638...
#> 5 1KME7376N10820    0   0   0       0 POLYGON ((-32.47984 -3.8549...
#> 6 1KME7376N10821    0   0   0       0 POLYGON ((-32.48054 -3.8459...
```

``` r
plot(st_geometry(aoi))
plot(st_geometry(aoi_grid), add = TRUE, border = "red")
```

<img src="man/figures/README-figure2-1.png" width="100%" />

The original Brazilian statistical grid was made available at the
following cell sizes: ‘500KM’, ‘100KM’, ‘50KM’, ‘10KM’, ‘5KM’, ‘1KM’ and
‘200M’. Using a different cell size value will result in an error. The
‘200M’ cell size grid covers only urban areas and is complemented by
cells with ‘1KM’ cell size. If you want to build a grid with ‘200M’ cell
size covering the entire area of interest (and without population census
data), you can use the function `gridbr_make()`.

Retrieving population census data requires either an internet connection
or the support package
[gridbr.data](https://github.com/luisfelipebr/gridbr.data). If the user
does not meet these requirements, the output will include only the
cell’s id and geometry, but not population census data.

If the user meet the requirements to retrieve population census data but
don’t want to include it, they can specify it by setting the optional
`census_data` parameter to FALSE.

``` r
aoi_grid <- gridbr_download(input = aoi,
                            cellsize = "1KM",
                            census_data = FALSE)
```

``` r
head(aoi_grid)
#> Simple feature collection with 6 features and 1 field
#> Geometry type: POLYGON
#> Dimension:     XY
#> Bounding box:  xmin: -32.47775 ymin: -3.890802 xmax: -32.42292 ymax: -3.877677
#> Geodetic CRS:  SIRGAS 2000
#>               id                           geom
#> 1 1KME7376N10816 POLYGON ((-32.47706 -3.8908...
#> 2 1KME7377N10816 POLYGON ((-32.46804 -3.8901...
#> 3 1KME7378N10816 POLYGON ((-32.45901 -3.8894...
#> 4 1KME7379N10816 POLYGON ((-32.44999 -3.8887...
#> 5 1KME7380N10816 POLYGON ((-32.44097 -3.8880...
#> 6 1KME7381N10816 POLYGON ((-32.43194 -3.8873...
```

There is another optional parameter related to the projection: by
default, the output will use the same CRS as the input. If you want to
keep the original grid CRS, with an equal area projection, you must set
the `equal_area` parameter to TRUE.

``` r
aoi_grid <- gridbr_download(input = aoi,
                            cellsize = "1KM",
                            equal_area = TRUE)
```

``` r
head(aoi_grid)
#> Simple feature collection with 6 features and 5 fields
#> Geometry type: POLYGON
#> Dimension:     XY
#> Bounding box:  xmin: 7376000 ymin: 10816000 xmax: 7377000 ymax: 10822000
#> CRS:           +proj=aea +lat_0=-12 +lon_0=-54 +lat_1=-2 +lat_2=-22 +x_0=5000000 +y_0=10000000 +ellps=GRS80 +units=m +no_defs
#>               id MASC FEM POP DOM_OCU                       geometry
#> 1 1KME7376N10816    0   0   0       0 POLYGON ((7376000 10816000,...
#> 2 1KME7376N10817    0   0   0       0 POLYGON ((7376000 10817000,...
#> 3 1KME7376N10818    0   0   0       0 POLYGON ((7376000 10818000,...
#> 4 1KME7376N10819    0   0   0       0 POLYGON ((7376000 10819000,...
#> 5 1KME7376N10820    0   0   0       0 POLYGON ((7376000 10820000,...
#> 6 1KME7376N10821    0   0   0       0 POLYGON ((7376000 10821000,...
```

### gridbr_make(input, cellsize)

**`gridbr_make()`** allows the user to make a standardized statistical
grid using any cell size (without population census data). An input
geospatial sf object and the cell size (in meters) must be specified.

``` r
aoi_grid <- gridbr_make(input = aoi,
                        cellsize = 100)
```

``` r
head(aoi_grid)
#> Simple feature collection with 6 features and 1 field
#> Geometry type: POLYGON
#> Dimension:     XY
#> Bounding box:  xmin: -32.47588 ymin: -3.883487 xmax: -32.4704 ymax: -3.882175
#> Geodetic CRS:  SIRGAS 2000
#>                     gid                           geom
#> 1 100ME7376250N10816850 POLYGON ((-32.47581 -3.8834...
#> 2 100ME7376350N10816850 POLYGON ((-32.47491 -3.8834...
#> 3 100ME7376450N10816850 POLYGON ((-32.47401 -3.8833...
#> 4 100ME7376550N10816850 POLYGON ((-32.4731 -3.88328...
#> 5 100ME7376650N10816850 POLYGON ((-32.4722 -3.88321...
#> 6 100ME7376750N10816850 POLYGON ((-32.4713 -3.88314...
```

``` r
plot(st_geometry(aoi))
plot(st_geometry(aoi_grid), add = TRUE, border = "red")
```

<img src="man/figures/README-figure3-1.png" width="100%" />

An alternative unique identifier **gid** and the cell’s geometry are
included in the output. If the cell size is contained in the original
statistical grid pool (500000, 100000, 50000, 10000, 5000, 1000, 200),
the original **id** is also included in the output.

There is an optional parameter related to the projection: by default,
the output will use the same CRS as the input. If you want to keep the
original grid CRS, with an equal area projection, you must set the
`equal_area` parameter to TRUE.

``` r
aoi_grid <- gridbr_make(input = aoi,
                        cellsize = 100,
                        equal_area = TRUE)
```

``` r
head(aoi_grid)
#> Simple feature collection with 6 features and 1 field
#> Geometry type: POLYGON
#> Dimension:     XY
#> Bounding box:  xmin: 7376200 ymin: 10816800 xmax: 7376800 ymax: 10816900
#> CRS:           +proj=aea +lat_0=-12 +lon_0=-54 +lat_1=-2 +lat_2=-22 +x_0=5000000 +y_0=10000000 +ellps=GRS80 +units=m +no_defs
#>                     gid                           geom
#> 1 100ME7376250N10816850 POLYGON ((7376200 10816800,...
#> 2 100ME7376350N10816850 POLYGON ((7376300 10816800,...
#> 3 100ME7376450N10816850 POLYGON ((7376400 10816800,...
#> 4 100ME7376550N10816850 POLYGON ((7376500 10816800,...
#> 5 100ME7376650N10816850 POLYGON ((7376600 10816800,...
#> 6 100ME7376750N10816850 POLYGON ((7376700 10816800,...
```


NEWS.md
# gridbr 

# gridbr v0.1.3 (2022-08-10)

* Updated README and documentation

# gridbr v0.1.2 (2022-07-25)

* Fixed 'gridbr_make' function id
* Added 'gridbr_make' documentation

# gridbr v0.1.1 (2022-07-22)

* Added 'gridbr_make' function
* Small fixes made to 'gridbr_download' function
* Small fixes made to the documentation

# gridbr v0.1.0 (2022-07-20)

* Release of 'gridbr' v0.1.0 on GitHub with the following function:
  * gridbr_download


R/gridbr_make.R
#' Make Brazilian statistical grid.
#'
#' Make Brazilian statistical grid covering (\code{input}).
#' @param input object of class sf, sfc or sfg.
#' @param cellsize integer. Target cell size (in meters). Must be greater than 1.
#' @param equal_area logical. Set to TRUE if you want to use the original grid CRS, with an equal area projection.
#' @return Brazilian statistical grid
#' @export

gridbr_make <- function(input, cellsize, equal_area = FALSE) {

  # error: input class
  if (!any(class(input) %in% c("sf", "sfc", "sfg"))) stop("input must be of class simple features (sf)")

  # error: cell size
  if (cellsize < 1 & cellsize%%1==0) stop('cell size must be an integer and greater than 1.')

  # import gridbr crs
  gridbr_crs <- gridbr::gridbr_crs

  # save input crs
  input_crs <- sf::st_crs(input)

  # change input crs
  input <- sf::st_transform(input, crs = gridbr_crs)

  # error: input coverage
  if (sf::st_bbox(input)[[1]] < 2800000 |
      sf::st_bbox(input)[[2]] < 7350000 |
      sf::st_bbox(input)[[3]] > 2800000 + (10 * 500000) |
      sf::st_bbox(input)[[4]] > 7350000 + (10 * 500000)) {
    stop("the input is outside IBGE Statistical Grid coverage.")
  }

  # get bbox of simple features
  input <- sf::st_bbox(input)

  # fix bbox to match IBGE grid
  input[1] <- 2800000 + (floor((input[1] - 2800000) / cellsize) * cellsize)
  input[2] <- 7350000 + (floor((input[2] - 7350000) / cellsize) * cellsize)
  input[3] <- 2800000 + (ceiling((input[3] - 2800000) / cellsize) * cellsize)
  input[4] <- 7350000 + (ceiling((input[4] - 7350000) / cellsize) * cellsize)

  # bbox class to sf class
  input <- sf::st_as_sfc(input)

  # make the grid
  grid <- sf::st_make_grid(
    x = input,
    cellsize = cellsize,
    crs = gridbr_crs,
    what = "polygons",
    square = TRUE,
    flat_topped = FALSE
  )

  # create a support dataset, to create id
  gridc <- sf::st_coordinates(suppressWarnings(sf::st_centroid(grid)))

  # grid class to sf (again)
  grid <- sf::st_as_sf(grid)

  # rename geometry to geom
  sf::st_geometry(grid) <- "geom"

  grid$gid <- paste0(
    ifelse(cellsize >= 1000, cellsize/1000, cellsize),
    ifelse(cellsize >= 1000, "KM", "M"),
    "E",
    floor(gridc[, 1]),
    "N",
    floor(gridc[, 2])
  )

  if (cellsize %in% c(500000, 100000, 50000, 10000, 5000, 1000)) {

    # recreate IBGE id
    grid$id <- paste0(
      cellsize/1000,
      "KM",
      "E",
      as.numeric(substr(formatC((2800000 + (floor((gridc[, 1] - 2800000) / cellsize) * cellsize)), width = 7, format = "d", flag = "0"), start = 1, stop = 4)),
      "N",
      as.numeric(substr(formatC((7350000 + (floor((gridc[, 2] - 7350000) / cellsize) * cellsize)), width = 8, format = "d", flag = "0"), start = 1, stop = 5))
    )

  }

  if(cellsize == 200) {

    # recreate IBGE id
    grid$id <- paste0(
      "200M",
      "E",
      as.numeric(substr(formatC((2800000 + (floor((gridc[, 1] - 2800000) / 200) * 200)), width = 7, format = "d", flag = "0"), start = 1, stop = 5)),
      "N",
      as.numeric(substr(formatC((7350000 + (ceiling((gridc[, 2] - 7350000) / 200) * 200)), width = 8, format = "d", flag = "0"), start = 1, stop = 6))
    )

  }

  grid <- rbind(grid)

  ##### equal_area parameter ----
  if (equal_area == FALSE) {
    grid <- sf::st_transform(grid, input_crs)
  }

  ##### export final object ----
  return(grid)

}


R/gridbr_download.R
#' Download Brazilian statistical grid with population census data.
#'
#' Make Brazilian statistical grid covering (\code{input}) with 2010 population census data.
#' @param input object of class sf, sfc or sfg.
#' @param cellsize character. Target cell size. Must be one of: "500KM", "100KM", "50KM", "10KM", "5KM", "1KM" and "200M".
#' @param census_data logical. Set to FALSE if you don't want to include 2010 population census data.
#' @param equal_area logical. Set to TRUE if you want to use the original grid CRS, with an equal area projection.
#' @return Brazilian statistical grid
#' @export

gridbr_download <- function(input, cellsize, census_data = TRUE, equal_area = FALSE) {

  # error: input class
  if (!any(class(input) %in% c("sf", "sfc", "sfg"))) stop("input must be of class simple features (sf)")

  # message: internet connection
  if (suppressWarnings(tryCatch(
    {
      readLines("http://example.com/", n = 1)
    },
    error = function(e) FALSE
  )) == FALSE & "gridbr.data" %in% utils::installed.packages() == FALSE) {
    message("couldn't connect to the internet. Population census data will not be loaded.")
    census_data <- FALSE
  }

  # message: gridbr.data
  if ("gridbr.data" %in% utils::installed.packages() == FALSE) message("population census data can be loaded faster by installing the package 'gridbr.data'. See https://www.github.com/luisfelipebr/gridbr.data")

  # create cell sizes support vector
  cellsizes <- c("200M", "1KM", "5KM", "10KM", "50KM", "100KM", "500KM")

  # error: cell size
  if (!cellsize %in% cellsizes) stop('cell size must be equal to one of the following values: "500KM", "100KM", "50KM", "10KM", "5KM", "1KM" or "200M"')

  # import gridbr crs
  gridbr_crs <- gridbr::gridbr_crs

  # save input crs
  input_crs <- sf::st_crs(input)

  ##### MAIN FUNCTION - BUILD GRID ----

  # change input crs
  input <- sf::st_transform(input, crs = gridbr_crs)

  # error: input coverage
  if (sf::st_bbox(input)[[1]] < 2800000 |
    sf::st_bbox(input)[[2]] < 7350000 |
    sf::st_bbox(input)[[3]] > 2800000 + (10 * 500000) |
    sf::st_bbox(input)[[4]] > 7350000 + (10 * 500000)) {
    stop("the input is outside IBGE Statistical Grid coverage.")
  }

  if (cellsize == "200M") {
    ##### 200M cellsize rule ----

    ##### PART 1 - 200M ----

    # get bbox of simple features
    input1 <- sf::st_bbox(input)

    # standardize bbox
    input1[1] <- 2800000 + (floor((input1[1] - 2800000) / 1000) * 1000)
    input1[2] <- 7350000 + (floor((input1[2] - 7350000) / 1000) * 1000)
    input1[3] <- 2800000 + (ceiling((input1[3] - 2800000) / 1000) * 1000)
    input1[4] <- 7350000 + (ceiling((input1[4] - 7350000) / 1000) * 1000)

    # bbox class to sf class
    input1 <- sf::st_as_sfc(input1)

    # make the grid
    grid1 <- sf::st_make_grid(
      x = input1,
      cellsize = 200,
      crs = gridbr_crs,
      what = "polygons",
      square = TRUE,
      flat_topped = FALSE
    )

    # create a support dataset, to create id
    grid1c <- sf::st_coordinates(suppressWarnings(sf::st_centroid(grid1)))

    # grid class to sf (again)
    grid1 <- sf::st_as_sf(grid1)

    # rename geometry to geom
    sf::st_geometry(grid1) <- "geom"

    # recreate the IBGE id
    grid1$id <- paste0(
      "200M",
      "E",
      as.numeric(substr(formatC((2800000 + (floor((grid1c[, 1] - 2800000) / 200) * 200)), width = 7, format = "d", flag = "0"), start = 1, stop = 5)),
      "N",
      as.numeric(substr(formatC((7350000 + (ceiling((grid1c[, 2] - 7350000) / 200) * 200)), width = 8, format = "d", flag = "0"), start = 1, stop = 6))
    )

    grid1 <- grid1[grid1$id %in% gridbr::gridbr_urban, ]

    ##### PART 2 - 1KM ----

    # get bbox of simple features
    input2 <- sf::st_bbox(input)

    # fix bbox to match IBGE grid
    input2[1] <- 2800000 + (floor((input2[1] - 2800000) / 1000) * 1000)
    input2[2] <- 7350000 + (floor((input2[2] - 7350000) / 1000) * 1000)
    input2[3] <- 2800000 + (ceiling((input2[3] - 2800000) / 1000) * 1000)
    input2[4] <- 7350000 + (ceiling((input2[4] - 7350000) / 1000) * 1000)

    # bbox class to sf class
    input2 <- sf::st_as_sfc(input2)

    # make the grid
    grid2 <- sf::st_make_grid(
      x = input2,
      cellsize = 1000,
      crs = gridbr_crs,
      what = "polygons",
      square = TRUE,
      flat_topped = FALSE
    )

    # grid class to sf (again)
    grid2 <- sf::st_as_sf(grid2)

    # rename geometry to geom
    sf::st_geometry(grid2) <- "geom"

    # get only outside intersection
    inters <- lengths(sf::st_intersects(suppressWarnings(sf::st_centroid(grid2)), suppressWarnings(sf::st_centroid(grid1)))) > 0
    grid2 <- grid2[!inters, ]

    # create a support dataset, to create id
    grid2c <- sf::st_coordinates(suppressWarnings(sf::st_centroid(grid2)))

    # recreate the IBGE id
    grid2$id <- paste0(
      "1KM",
      "E",
      as.numeric(substr(formatC((2800000 + (floor((grid2c[, 1] - 2800000) / 1000) * 1000)), width = 7, format = "d", flag = "0"), start = 1, stop = 4)),
      "N",
      as.numeric(substr(formatC((7350000 + (floor((grid2c[, 2] - 7350000) / 1000) * 1000)), width = 8, format = "d", flag = "0"), start = 1, stop = 5))
    )

    grid <- rbind(grid1, grid2)
  } else {

    #### KM cellsize rule ----

    # create support cell size object (numeric)
    cellsize2 <- as.numeric(gsub(
      pattern = "KM",
      replacement = "000",
      x = cellsize
    ))

    # get bbox of simple features
    input <- sf::st_bbox(input)

    # fix bbox to match IBGE grid
    input[1] <- 2800000 + (floor((input[1] - 2800000) / cellsize2) * cellsize2)
    input[2] <- 7350000 + (floor((input[2] - 7350000) / cellsize2) * cellsize2)
    input[3] <- 2800000 + (ceiling((input[3] - 2800000) / cellsize2) * cellsize2)
    input[4] <- 7350000 + (ceiling((input[4] - 7350000) / cellsize2) * cellsize2)

    # bbox class to sf class
    input <- sf::st_as_sfc(input)

    # make the grid
    grid <- sf::st_make_grid(
      x = input,
      cellsize = cellsize2,
      crs = gridbr_crs,
      what = "polygons",
      square = TRUE,
      flat_topped = FALSE
    )

    # create a support dataset, to create id
    gridc <- sf::st_coordinates(suppressWarnings(sf::st_centroid(grid)))

    # grid class to sf (again)
    grid <- sf::st_as_sf(grid)

    # rename geometry to geom
    sf::st_geometry(grid) <- "geom"

    # recreate IBGE id
    grid$id <- paste0(
      cellsize,
        "E",
        as.numeric(substr(formatC((2800000 + (floor((gridc[, 1] - 2800000) / cellsize2) * cellsize2)), width = 7, format = "d", flag = "0"), start = 1, stop = 4)),
        "N",
        as.numeric(substr(formatC((7350000 + (floor((gridc[, 2] - 7350000) / cellsize2) * cellsize2)), width = 8, format = "d", flag = "0"), start = 1, stop = 5))
      )

    grid <- rbind(grid)
  }

  ##### census_data parameter ----
  if (census_data == TRUE) {
    j <- 1

    data <- list()

    for (i in cellsizes) {
      cs <- any(grepl(pattern = i, x = grid$id))

      if (cs == TRUE) {

        # download or get from package, IBGE census data
        if (requireNamespace("gridbr.data", quietly = TRUE)) {
          assign(paste0("gridbr_", i), eval(parse(text = paste0("gridbr.data::gridbr_", i))))
        } else {
          utils::download.file(paste0("https://github.com/luisfelipebr/gridbr.data/raw/master/data/gridbr_", i, ".rda"), "inst/Temp")
          load("inst/Temp")
        }

        data[j] <- mget(paste0("gridbr_", i))

        j <- j + 1
      }
    }

    data <- do.call(rbind, data)

    # merge data with grid
    grid <- base::merge(grid, data, by = "id", all.x = TRUE)

    # remove na's
    grid$MASC <- ifelse(is.na(grid$MASC), 0, grid$MASC)
    grid$FEM <- ifelse(is.na(grid$FEM), 0, grid$FEM)
    grid$POP <- ifelse(is.na(grid$POP), 0, grid$POP)
    grid$DOM_OCU <- ifelse(is.na(grid$DOM_OCU), 0, grid$DOM_OCU)
  }

  ##### equal_area parameter ----
  if (equal_area == FALSE) {
    grid <- sf::st_transform(grid, input_crs)
  }

  ##### export final object ----
  return(grid)
}
